lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1698,65 +1755,172 @@ ArrayAttr ClassOp::getPortAnnotationsAttr() {
 
 hw::ModulePortInfo ClassOp::getPortList() { return ::getPortListImpl(*this); }
 
-LogicalResult
-ClassOp::verifyType(ClassType type,
-                    function_ref<InFlightDiagnostic()> emitError) {
-  // This check is probably not required, but done for sanity.
-  auto name = type.getNameAttr().getAttr();
-  auto expectedName = getModuleNameAttr();
-  if (name != expectedName)
-    return emitError() << "type has wrong name, got " << name << ", expected "
-                       << expectedName;
+BlockArgument ClassOp::getArgument(size_t portNumber) {
+  return getBodyBlock()->getArgument(portNumber);
+}
 
-  auto elements = type.getElements();
-  auto numElements = elements.size();
-  auto expectedNumElements = getNumPorts();
-  if (numElements != expectedNumElements)
-    return emitError() << "has wrong number of ports, got " << numElements
-                       << ", expected " << expectedNumElements;
+bool ClassOp::canDiscardOnUseEmpty() {
+  // ClassOps are referenced by ClassTypes, and these uses are not
+  // discovereable by the symbol infrastructure. Return false here to prevent
+  // passes like symbolDCE from removing our classes.
+  return false;
+}
 
-  for (unsigned i = 0; i < numElements; ++i) {
-    auto element = elements[i];
+//===----------------------------------------------------------------------===//
+// ExtClassOp
+//===----------------------------------------------------------------------===//
 
-    auto name = element.name;
-    auto expectedName = getPortNameAttr(i);
-    if (name != expectedName)
-      return emitError() << "port #" << i << " has wrong name, got " << name
-                         << ", expected " << expectedName;
+void ExtClassOp::build(OpBuilder &builder, OperationState &result,
+                       StringAttr name, ArrayRef<PortInfo> ports) {
+  assert(
+      llvm::all_of(ports,
+                   [](const auto &port) { return port.annotations.empty(); }) &&
+      "class ports may not have annotations");
 
-    auto direction = element.direction;
-    auto expectedDirection = getPortDirection(i);
-    if (direction != expectedDirection)
-      return emitError() << "port " << name << " has wrong direction, got "
-                         << direction::toString(direction) << ", expected "
-                         << direction::toString(expectedDirection);
+  buildModuleWithoutAnnos(builder, result, name, ports);
+}
 
-    auto type = element.type;
-    auto expectedType = getPortType(i);
-    if (type != expectedType)
-      return emitError() << "port " << name << " has wrong type, got " << type
-                         << ", expected " << expectedType;
+void ExtClassOp::print(OpAsmPrinter &p) {
+  p << " ";
+
+  // Print the visibility of the class.
+  StringRef visibilityAttrName = SymbolTable::getVisibilityAttrName();
+  if (auto visibility = (*this)->getAttrOfType<StringAttr>(visibilityAttrName))
+    p << visibility.getValue() << ' ';
```

Comment:
I don't care which is used, but within same function please be consistent how a single space is sent out (char vs single-char string literal) :upside_down_face: .